### PR TITLE
Fix: Router white flash on first navigation to page

### PR DESCRIPTION
**Pull request by Tobbe**
This is an alternative attempt at a solution for #2124 

When I was reviewing @callingmedic911's PR for the same issue (#3767) I did some refactoring while getting myself up-to-speed on the code again. And while doing so (and after reading @jtoar's excellent link to the React beta docs) I came up with a (in my eyes) simpler solution to the problem.

This PR has two commits (as of writing this). The first one is just the refactoring I did. The second one is the actual solution to the linked issue.

The TL;DR of it is: Set a hard-coded `key` on the `<Route>`s so that `<PageLoader>` can keep its state while navigating.

Need to update the docs: https://redwoodjs.com/docs/prerender#flash-after-page-load

### CODE REVIEW

**Comment by callingmedic911** [comment]
@Tobbe I checked-out this locally, this is a neat solution! But I'm afraid unmounting could still happen if navigation is not between the `Route` to `Route`. It could be from the old `Route` to a `Set` which contains the new route. Here's a small repoduction:
```jsx
<Router>
  <Set wrap={TestWrapper}>
    <Route path="/second" page={SecondPage} name="second" />
  </Set>
  <Route path="/first" page={FirstPage} name="first" />
  <Route notfound page={NotFoundPage} />
</Router>
```
From what I understood, if it was only possible to wrap Set once, even we could have handled it with simple control flow but Set could be multi-level nested. Can we somehow retain the route state even if parents change?

**Comment by Tobbe** [comment]
> It could be from the old `Route` to a `Set` which contains the new route

Yes, I considered this too.

Here's what I'm afraid will happen if we have a "global" Page. But I haven't fully verified.

The PageLoader is only for the actual page, not the Set and the wrappers around it.
So if you jump from a page in `BlogLayout` to a page in `AdminLayout` you'd get the old page rendered inside `AdminLayout` until `pageLoadingDelay` ms has passed, or the new page has finished downloading. I thought this would be weird, so I made sure everything was unmounted as soon as you "leave" the current Set

**Comment by callingmedic911** [comment]
> The PageLoader is only for the actual page, not the Set and the wrappers around it. So if you jump from a page in `BlogLayout` to a page in `AdminLayout` you'd get the old page rendered inside `AdminLayout` until `pageLoadingDelay` ms has passed, or the new page has finished downloading. I thought this would be weird, so I made sure everything was unmounted as soon as you "leave" the current Set

You're right, that would be a very inconsistent state. Worse than the white screen IMO.

>Yes, I considered this too.

I see the white page for this case, am I missing something here?

**Comment by Tobbe** [comment]
> I see the white page for this case, am I missing something here?

Sorry, English is difficult... I just meant I knew about the scenario you mentioned. And that I thought the white flash was better than the alternative I described.

**Comment by callingmedic911** [comment]
Oh, I see. Would love to know your thought on this: https://github.com/redwoodjs/redwood/pull/3767#issuecomment-974786065
Specifically:
>Long term fix can be a bit of refactoring. Theoretically, we have to put Loader outside of Route/activeChildren and pass download page module as a prop to the current Route. Then instead of returning PageLoader from InternalRoute, it can return Page directly. But this seemed too much code change considering we are very near to v1. However, if we agree to it I can proceed with guidance from you guys - @Tobbe @jtoar.

Having PageLoader outside will hold the "swapping" process (including Sets) until the page is ready. What do you think? Too much change?

**Comment by Tobbe** [comment]
> Having PageLoader outside will hold the "swapping" process (including Sets) until the page is ready. What do you think? Too much change?

I felt I couldn't get a good grip on how much work this would be, so I just got started on it. And it doesn't look like it'll be too bad. I'll spend a few hours more on it tomorrow to finish it off.

**Comment by Tobbe** [comment]
> it doesn't look like it'll be too bad. I'll spend a few hours more on it tomorrow to finish it off.

Hah! That wasn't true. This turned out to be a much bigger and complicated change than I first thought. 

---

@jtoar This touches on the a11y stuff you've built. Had to update the Set test snapshots. Can you please take a look?
@dac09 pre-rendering died due to my changes. I've gotten them back to life, but please do take a look. How much confidence do you have in our pre-render tests?

**Comment by dac09** [comment]
Prerender tests are pretty solid!

**Comment by Tobbe** [comment]
Awesome! That's a relief :slightly_smiling_face:

**Comment by Tobbe** [comment]
@callingmedic911 I'd love your review of this 🙏

**Comment by callingmedic911** [comment]
> @callingmedic911 I'd love your review of this 🙏

Glad to. Huge props to you for pulling this big of a change this fast. 🙌 🚀 I'll try to squeeze some time this weekend to review.

File: packages/router/src/active-route-loader.tsx [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    }
+  }
+
+  useEffect(() => {
+    mounted.current = true
+
+    return () => {
+      mounted.current = false
+    }
+  }, [])
```
Comment (jtoar) [inline packages/router/src/active-route-loader.tsx:116]: I feel like we could get a few readability wins if we start extracting some of the logic here into hooks. For this one, something like:

```js
const useIsMounted = () => {
  const isMounted = useRef(false)

  useEffect(() => {
    isMounted.value = true
    return () => {
      isMounted.value = false
    }
  }, [])

  return isMounted.value
}
```

File: packages/router/src/active-route-loader.tsx [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+      waitingFor.current = name
+      const module = await loader()
+
+      // Remove the timeout because the page has loaded.
+      clearLoadingTimeout()
+
+      // Only update all state if we're still interested (i.e. we're still
+      // waiting for the page that just finished loading)
+      if (mounted.current && name === waitingFor.current) {
+        unstable_batchedUpdates(() => {
```
Comment (jtoar) [inline packages/router/src/active-route-loader.tsx:175]: Just noting that this is stable in react 18:
- https://github.com/reactwg/react-18/discussions/21

**Review by jtoar — COMMENTED** [review]
Awesome job @Tobbe! This one's a doozy. I'll have to take more than one pass at reviewing it.

First off, super happy that `ActiveRouteLoader` is a function component! When I was digging into React 18 I was worried that it might take a lot of nontrivial work to get the Router concurrent-mode/suspense ready. There's still some work to do but I'm feeling a lot better about it now that it's a function component.

I've tried this PR out locally on an app I have. The behavior's definitely fixed, and I haven't caught any regressions yet.

After this PR, I think we need to take some time to organize some of the code in the router package. For example, path-validation logic seems to be duplicated in a few places; it'd be easier to follow if it was centralized. Moreover, we need a legit "understanding the router" or contributing guide.

File: packages/router/src/active-route-loader.tsx [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+          state: 'PRE_SHOW',
+        },
+      }))
+
+      // Update the context if importing the page is taking longer
+      // than `delay`.
+      // Consumers of the context can show a loading indicator
+      // to signal to the user that something is happening.
+      loadingTimeout.current = setTimeout(() => {
+        unstable_batchedUpdates(() => {
```
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx:151]: 👍

File: packages/router/src/active-route-loader.tsx [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+  delay,
+  params,
+  whileLoadingPage,
+  children,
+}: Props) => {
+  const location = useLocation()
+  const [pageName, setPageName] = useState('')
+  const loadingTimeout = useRef<NodeJS.Timeout>()
+  const announcementRef = useRef<HTMLDivElement>(null)
+  const mounted = useRef(true)
```
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx:91]: In this case it doesn't matter if it starts out being `true` or `false`.
But it's an interesting question nonetheless 🙂 

Let's say we change it to be `const mounted = useRef(false)`.
Then, further down, we set it to `true` inside a useEffect with an empty dependency array.

useEffect hooks are called **after** render. So `mounted.current` would still have been `false` on the first render. But clearly the component is mounted, otherwise it couldn't have been rendered, right? You can't render an unmounted component, right?

On the other hand, people use this to simulate `componentWillMount()` in function components like this:

```
const willMount = useRef(true)

if (willMount.current) {
  // Before mounting
}

willMount.current = false
```

So, all this to say I'm not 100% sure exactly when a function component is considered mounted.

File: packages/router/src/active-route-loader.tsx (hunk 1 above)
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx:116]: You mean `isMounted.current`, right? (Not `value`)

File: packages/router/src/active-route-loader.tsx [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    // override the types for this condition
+    const syncPageLoader = spec.loader as unknown as synchronousLoaderSpec
+    const PageFromLoader = syncPageLoader().default
+
+    const prerenderLoadingState: LoadingStateRecord = {
+      [path]: {
+        state: 'DONE',
+        page: PageFromLoader,
+      },
+    }
```
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx:207]: Done. Please lmk what you think

File: packages/router/src/active-route-loader.tsx (hunk 1 above)
Comment (jtoar) [reply packages/router/src/active-route-loader.tsx]: Right! I was doing some vue stuff a bit the other day and I think they have refs with `value` properties.

File: packages/router/src/active-route-loader.tsx [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+  const [loadingState, setLoadingState] = useState<LoadingStateRecord>({
+    [path]: { page: ArlNullPage, state: 'PRE_SHOW' },
+  })
+  const [renderedChildren, setRenderedChildren] = useState<
+    React.ReactNode | undefined
+  >(children)
+  const [renderedPath, setRenderedPath] = useState(path)
+  const [renderedLocation, setRenderedLocation] = useState({ ...location })
+
+  loadingState[path] = loadingState[path] || 'PRE_SHOW'
```
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx]: Yeah, looking at this again this was a very weird solution. I removed the line and solved the problem in a much nicer way. Thanks for pointing this out.

File: packages/router/src/active-route-loader.tsx (hunk 1 above)
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx]: Done

Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx]: I chose to put the hook in a separate file right next to this file rather than in some more globally accessible place because it's not very often you have to do something like this. Usually there are better way to handle your problems. So didn't want to make it too easy to reuse this hook 🙂

File: packages/router/src/active-route-loader.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx]: Let's say you have this React app

```tsx
import React, { useEffect, useRef } from 'react';
import { render } from 'react-dom';

const App = () => {
  const mounted = useRef(false)

  useEffect(() => {
    mounted.current = true

    return () => {
      mounted.current = false
    }
  }, [])

  return <p>mounted: {'' + mounted.current}</p>
}

render(<App />, document.querySelector('#app'));
```

It will render once, and it'll say "mounted: false". That feels a little weird to me 🙂

File: packages/router/src/active-route-loader.tsx (hunk 1 above)
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx]: Using your hook looks like this, right?
`const isMounted = useIsMounted()`

Now, if you use `isMounted` inside a `useEffect` you'll have to put `isMounted` in the dependency array. And that means when `isMounted` changes it'll trigger a rerender. And we don't want refs to trigger rerenders.

So what we have to do is return a function instead: `return useCallback(() => isMounted.current, [])`. Now `isMounted` is a stable value that's safe to put in dependency arrays. Just also have to change the usage to `isMounted()`

**Comment by Tobbe** [comment]
The CI jobs keeps failing to install. Not sure how to get this unstuck

**Review by callingmedic911 — COMMENTED** [review]
Looking good!

**Comment by thedavidprice** [comment]
It would be great to get this into the next release. How are you feeling about it @Tobbe (and how are you feeling about your availability)?

File: packages/router/src/active-route-loader.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx:207]: @jtoar Have you had a chance to look at this yet?

File: packages/router/src/active-route-loader.tsx [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+  | undefined
+>
+
+interface ActivePageState {
+  loadingState: LoadingStateRecord
+}
+
+const ActivePageContext = React.createContext<ActivePageState | undefined>(
+  undefined
+)
```
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/router/src/active-route-loader.tsx]: Extracted to its own file

File: packages/router/src/router.tsx [hunk 8]
```diff
@@ -69,10 +71,10 @@ const InternalRoute: React.VFC<InternalRouteProps> = ({
   name,
   redirect,
   notfound,
-  whileLoadingPage,
 }) => {
   const location = useLocation()
```
[the remark this answers is not in the document]
Comment (Tobbe) [reply packages/router/src/router.tsx:76]: As you mention, we're passing the correct location to ParamsProvider, so as long as the pages are using `useParams` there should be no mismatch in route and search params.

`useLocation` is a bigger issue. I wrote an issue about it some time ago https://github.com/redwoodjs/redwood/issues/3779

So a proper fix is a much bigger change. 

> 1. Unnecessary render as soon location changes.
> 2. Inconsistent state where `path` is new but `location` is the old one

1. Is it alright with you if we live with the extra render until we've thought more about how we want to tackle #3779?
2. Params are already taken care of as long as you use `useParams`. And `useLocation` is also talked about in the issue linked above. Is there anything else you're worried about?

**Comment by Tobbe** [comment]
Whatever I do I can't get rid of this error
image

I just now tried removing all my changes and just updating a comment. So pretty much running latest `main` and still get the error.

I've also tried reseting my branch to old commits like the v0.39 and v0.39.2 commits, and get the same error there as well.

I'm out of ideas for the moment.

**Comment by jtoar** [comment]
I think I’ve isolated this to `cross-undici-fetch` upgrading from under us from `0.1.4` to `0.1.10`. Here’s a repo with the reproduction: https://github.com/jtoar/test-web-repro. The last commit in it is patching the version of `cross-undici-fetch`. `yarn rw test web` works on this commit, but if you revert to the one before it, it fails (like what's currently happening in CI).

**Review by jtoar — APPROVED** [review]
This looks good to go! I feel like this really opens the door to even more refactoring, like you brought up in your page loader issue (I think we could consolidate “validatePaths” too—I’ll make a few follow-up issues myself).

I’m really impressed by all the work you did with the router tests. I thought that we might have to be perennially on-the-look-out for the flash, but you’ve seriously tested it every which way. I’ll try to take a page out of your book when writing tests going forward.

I’ll leave you to merge 🚀
